### DISCUSSION TITLE
How do I embed a .res file into a PE file?

### DISCUSSION THREAD

**Discussion by jgh07** [dotnet/runtime#105271] [Q&A] [answered]
I'm using the new .NET 9 ``PersistedAssemblyBuilder`` to generate some assemblies and I want to embed version information from a .res file. In .NET Framework, I would have used ``AssemblyBuilder.DefineUnmanagedResource()`` for that.

I believe the equivalent for .NET 9 is the ``nativeResources`` parameter of the constructor of ``ManagedPEBuilder``. The documentation of ``ResourceSectionBuilder`` is severely lacking, so I didn't know how it works exactly, but I created this class:

````csharp
internal class VersionInfoBuilder : ResourceSectionBuilder
{
    public VersionInfoBuilder(string resFilePath)
    {
        if (resFilePath == null)
            return;

        rawBytes = File.ReadAllBytes(resFilePath);
    }

    private readonly byte[] rawBytes = null;

    protected override void Serialize(BlobBuilder builder, SectionLocation location)
    {
        if (rawBytes == null)
            return;

        builder.WriteBytes(rawBytes);
    }
}
````
I'm using it like this:
````csharp
public static ManagedPEBuilder CreatePEBuilder(MethodInfo entryPoint, string resFilePath = null)
{
    MetadataBuilder mb = Context.Assembly.GenerateMetadata(out BlobBuilder ilStream, out BlobBuilder mappedFieldData);
    PEHeaderBuilder headerBuilder = new(
        imageBase: 0x00400000,
        imageCharacteristics: Characteristics.ExecutableImage);

    ManagedPEBuilder peBuilder = new(
        headerBuilder,
        new(mb),
        ilStream,
        mappedFieldData,
        entryPoint: MetadataTokens.MethodDefinitionHandle(entryPoint.MetadataToken),
        nativeResources: new VersionInfoBuilder(resFilePath));

    return peBuilder;
}
````

It does not work. Can anyone tell me how to correctly do it?

**Comment by jgh07** [comment]
Maybe @buyaa-n or @jkotas or one of the other guys responsible for the .NET 9 implementation of ``PersistedAssemblyBuilder`` can look over this, if you have time.

**Reply by jkotas** [reply]
Does https://learn.microsoft.com/en-us/dotnet/fundamentals/runtime-libraries/system-reflection-emit-persistedassemblybuilder#add-resources-with-persistedassemblybuilder help?

**Reply by jgh07** [reply]
That seems to be adding managed resources? I want to add an unmanaged, compiled .res file.

I'm using 7-Zip to inspect the generated assemblies and in my case, 7-Zip shows me a file called .rsrc after I generate an assembly using my above code.

Any other assembly containing version information is instead shown containing a folder called .rsrc with a file called version.txt. I don't think this is accurate with the actual PE structure, but it is a good visualization.

**Comment by MichalStrehovsky** [comment, marked as answer]
RES file format and contents of the .rsrc section are similar, but subtly different. https://github.com/dotnet/corert/blob/master/src/ILCompiler.Build.Tasks/src/DumpNativeResources.cs is a sample converter from `.rsrc` section to a res file. You'll need to do this in the opposite direction.

When doing the conversion, you'll see that the data structures within the rsrc section refer to offsets from the beginning of the file (not from the beginning of the section). The `SectionLocation` argument that gets passed to your `Serialize` override should come handy for that since it tells where in the file rsrc section is getting placed.

**Reply by jgh07** [reply]
Thank you! It was quite a challenge, but I managed to get it working.
